documentation/Tracer.md

```diff
@@ -0,0 +1,201 @@
+supported instructions.
+
+
+
+When one of these instruction is encountered the trace will indicate
+and on or off event using '+' or '-' before the instruction mnemonic.
+
+Example:
+
+// tracing starts
+RevCPU[cpu0:InstTrace:18156000]: Core 0; Hart 0; Thread 1]; *I 0x10330:00301013  + slli    zero, zero, 3	 0x0<-zero zero<-0x0 pc<-0x10334 
+RevCPU[cpu0:InstTrace:18157000]: Core 0; Hart 0; Thread 1]; *I 0x10334:fec42703    lw      a4, -20(s0)	 0x3ffffba0<-s0 0x000003de<-[0x3ffffb8c,4] a4<-0x3de 
+RevCPU[cpu0:InstTrace:18158000]: Core 0; Hart 0; Thread 1]; *I 0x10338:3de00793    li      a5, 990	 0x0<-zero a5<-0x3de 
+RevCPU[cpu0:InstTrace:18159000]: Core 0; Hart 0; Thread 1]; *I 0x1033c:00f70463    beq     a4, a5, pc + 8	 0x3de<-a4 0x3de<-a5 pc<-0x10344 
+RevCPU[cpu0:InstTrace:18160000]: Core 0; Hart 0; Thread 1]; *I 0x10344:00201013  - slli    zero, zero, 2	 0x0<-zero zero<-0x0
+// no tracing between cycles 18160 and 27090
+RevCPU[cpu0:InstTrace:27090000]: Core 0; Hart 0; Thread 1]; *I 0x10358:00301013  + slli    zero, zero, 3	 0x0<-zero zero<-0x0 pc<-0x1035c 
+RevCPU[cpu0:InstTrace:27091000]: Core 0; Hart 0; Thread 1]; *I 0x1035c:fec42783    lw      a5, -20(s0)	 0x3ffffba0<-s0 0x00000000<-[0x3ffffb8c,4] a5<-0x0 
+RevCPU[cpu0:InstTrace:27092000]: Core 0; Hart 0; Thread 1]; *I 0x10360:00078463    beqz    a5, pc + 8	 0x0<-a5 0x0<-zero pc<-0x10368 
+RevCPU[cpu0:InstTrace:27093000]: Core 0; Hart 0; Thread 1]; *I 0x10368:00201013  - slli    zero, zero, 2	 0x0<-zero zero<-0x0 
+// tracing ends
+
+4.  Tracer Output
+-------------------
+
+4.1 Logging prefix
+-------------------
+
+  The tracer output is prefixed by the standard logging information with
+  an additional text to support extraction from the log file.
+
+   |--   Standard logging prefix                           --| key |
+    RevCPU[cpu0:InstTrace:18156000]: Core 0; Hart 0; Thread 1]; *I 
+
+4.2 Instruction Disassembly
+--------------------------
+
+4.2.1 REV_USE_SPIKE=ON
+----------------------
+
+  |  PC  : INSN      |  Disassembly     |
+  0x10474:00c50533    add     a0, a0, a2
+
+4.2.2 REV_USE_SPIKE=OFF
+-----------------------
+
+  |  PC  : INSN      |  Disassembly      |
+  0x10474:00c50533    add %rd, %rs1, %rs2
+
+4.3 Register to Register Format
+--------------------------------
+
+  |  PC   : INSN     |  Disassembly             |         Effects          |
+   0x10474:00c50533    add     a0, a0, a2	 0x14<-a0 0x50<-a2 a0<-0x64 
+
+  Notice that the effects include the same register naming conventions
+  as the disassembled instruction. 
+
+  Effects interpretation:
+   data<-reg  : Source register read
+   reg<-data  : Destination register write
+
+4.4 Register to Memory (Store)
+------------------------------
+  |  PC  : INSN    |     Disassembly    |
+  0x10220:fef42623  sw      a5, -20(s0)	 
+
+  |                    Effects                        |
+    0x3ffffb60<-s0 0x64<-a5 [0x3ffffb4c,4]<-0x00000064
+
+  Effects interpretation:
+   data<-reg  : Source register read
```

Comment:
This format is a bit confusing, in my opinion "reg<-data" would be better, or at least "data->reg"

Reply:
Perhaps a prefix indicator would be better? 

R(reg): data
W(data): reg 

At the very least I definitely agree with differentiating the direction of the arrows if we don't go with a prefix.

Reply:
I agree that arrows should go right to left, but I think the LHS of an arrow should always be a register or memory, and not the value being written.

For the value being written,
```
Reg <- Reg (= 0x12345678)
Mem <- Reg (= 0x12345678)
Reg <- 0x12345678
```
should be used, similar to how `gdb` prints values when examining variables / registers.

Reply:
I like the <- notation as is. It is overloaded, but conveys a lot of information succinctly. Keeping the arrows in one direction allows the focus to be on the values next to the arrows.